README.md
# Eta

Eta-2 Actis-class interceptor, sometimes referred to as the Jedi interceptor due to its popularity with Jedi pilots, was a Clone Wars-era Republic starfighter.

Think of Eta as a set of well crafted tasks that you can add to your Gulp workflow and help speed up your development.

For 40Digits, Eta serves as our build script for internal projects. The build script took much inspiration from [graypants'](https://github.com/greypants/gulp-starter) & [Chris Davies'](https://github.com/chrisdavies/gulp_starter_kit) starter kits. It was modified to work within the needs and requirements of 40Digits development.

Eta includes the following tools, tasks, and workflows:

- [Browserify](http://browserify.org/)
- [Sass](http://sass-lang.com/) (libsass with [source maps](https://github.com/sindresorhus/gulp-ruby-sass#sourcemap), and [autoprefixer](https://github.com/sindresorhus/gulp-autoprefixer))
- [Image optimization](https://www.npmjs.com/package/gulp-imagemin)
- [Sprite generation](https://www.npmjs.com/package/sprity)
- [Custom Icon Font generation](https://www.npmjs.com/package/gulp-iconfont)
- Error handling in the console [and in Notification Center](https://github.com/mikaelbr/gulp-notify)
- Compression task for production builds (CSS + JS)
- [BrowserSync](http://www.browsersync.io/)

## Requirements

- [Node.js](https://nodejs.org/download/)
- [Gulp](https://github.com/gulpjs/gulp)

***

Eta is made to work inside of any framework. Start your project however you'd like; Rails, Express, WordPress, etc. Then follow these steps.

### 1. Install Gulp

Eta can't do anything without Gulp. From the command line, navigate to the directory where you would like to run your gulp tasks.

```bash
npm install --save-dev gulp
```

### 2. Install Eta

```bash
npm install --save-dev gulp-eta
```

This adds Eta to your `node_modules` directory along with all of its dependencies. See [troubleshooting](https://github.com/40Digits/eta#troubleshooting) section if you run into errors.

### 3. Create `gulpfile.js`

In your current working directory, create a file called `gulpfile.js` which serves as the configuration file for Gulp. Here is where you link up Eta. Eta adds tasks to your Gulp module.

One great thing is that you aren't limited to what Eta provides! You can declare your own custom tasks, too.

Your `gulpfile` should look something like this:

```javascript
// dependencies
var gulp = require('gulp');
var eta = require('gulp-eta');

// instantiate eta passing in `gulp` as the first arg
// and then your configuration object as the second arg
eta(gulp, {
  // your options
});

// add some custom tasks if you want
gulp.task('mytask', function() {
  // blah blah blah
});
```

Checkout the [examples](/examples) for some common configurations.

### 4. Your choice

At this point you have three options.

1) You can either go ahead and run `gulp init` to use Eta's default configuration to setup your source files and install the necessary depencies.

2) You can pass in your own configuration to the Eta instance in `gulpfile.js` to meet your needs, then run `gulp init` to set up all of the source files where you need them.

3) You can skip `gulp init` and set up your project however you want. If you elect to go this route make sure to let Eta know where all of your source files are and where you want the compiled assets to be created using the `scaffold` option. Or else the Eta `gulp` tasks won't work.

### 5. Run the tasks

```bash
gulp
```
This will run the `default` gulp task, which has the following task dependencies: `['browserSync', 'symbols', 'sass', 'sprites', 'images', 'browserify']`.

[View all available tasks](#tasks)

***

## Scaffold

Eta uses the scaffold object to handle paths. Below is the default configuration. Override them to meet your needs. `source.root` and `assets.root` are relative to directory of your `gulpfile`, and all subsequent directories are respectively relative to `source.root` and `assets.root`.

If you want an `asset` to be in the root of your app (where your gulpfile lives) then set it to `'/'`.

*Note:* Make sure you set up your scaffold before you run `gulp init`.

**Defaults:**
```javascript
config.scaffold = {
  source: {
    root:     '_src',
    images:   'images',
    scripts:  'js',
    sprites:  'sprites',
    styles:   'sass',
    symbols:  'symbols',
    static:   'static'
  },
  assets: {
    root:     'assets',
    images:   'images',
    sprites:  'images/sprites',
    scripts:  'js',
    styles:   '/',
    symbols:  'fonts/symbols',
    static:   '/'
  }
}
```

The folder stucture looks like this:
```
├───_src
│   ├───images
│   ├───js
│   ├───sprites
│   ├───sass
│   ├───symbols
│   └───static
├───assets
│   ├───images
│   │   └───sprites
│   ├───js
│   └───fonts
│       └───symbols
├───gulpfile.js
├───style.css // styles go to app root by default
├───index.html // created from the static task
└───node_modules
```

**Examples:**

Checkout the [examples](/examples) for some common configurations.

If you need your assets folder to live in a `/public` folder:

```javascript
options.scaffold.assets.root = 'public/assets';
```

If you want to rename your source folder:

```javascript
options.scaffold.source.root = 'source';
```

If you want your CSS files in the assets folder:

```javascript
options.scaffold.assets.styles = 'css';
```

***

## Tasks

### `default`

Runs the specified default tasks.

**Options:**
```javascript
config.default = {
  tasks: ['browserSync', 'symbols', 'sass', 'sprites', 'images', 'browserify']
}
```

### `init`

Creates a `_src` directory, installs some starter modules. If you run this a second time it will abort if the `_src` directory has already been created.

Use `-p` or `--proxy` to set the name of the localhost for your project during initialization. Using the flag without any arguments will create a hostname using the name of the directory where your gulpfile is located prepended with "l.".

*Examples:*

```bash
gulp init -p
// uses l.DIRNAME for proxy

gulp init --proxy
// uses l.DIRNAME for proxy

gulp init -p super-cool.lh
// uses super-cool.lh for proxy

gulp init --proxy awesome.local
// uses awesome.local for proxy
```

*Note:* [Make sure your scaffold is configured](#scaffold) before you run this.

### `symbols`

Generates your icon font, preview file, and sass file.

**Options:**
```javascript
options.symbols = {
  settings: {
    fontName: 'symbols',
    appendCodepoints: false,
    centerHorizontally: true,
    normalize: true,
    fontHeight: false
  },
  renameSass: {
    basename: '_symbols',
    extname: '.scss'
  }
}
```

### `sass`

Compiles your sass files.

**Options:**
```javascript
options.sass = {
  // for all available options: https://github.com/sass/node-sass#options
  settings: {
    indentedSyntax: true,
    outputStyle: 'nested'
  },
  globbing: {
    extensions: ['.scss', '.sass']
  }
};

// for all available options: https://www.npmjs.com/package/gulp-autoprefixer#api
options.autoprefixer = {
  browsers: [
    'last 2 versions',
    'safari 5',
    'ie 8',
    'ie 9',
    'android 4'
  ],
  cascade: true
};
```

### `images`

Moves image copies from a source folder, performs optimizations, then outputs them into the assets folder.

**Options:**
```javascript
options.images = {
  // for all available options: https://www.npmjs.com/package/gulp-imagemin#api
  settings: {
    progressive: true,
    optimizationLevel: 4
  }
}
```

### `sprites`

Compiles sprite assets into a sprite sheet, and generates a sass file for mixins & variable use.

**Options:**
```javascript
options.sprites = {
  // for all available options: https://www.npmjs.com/package/sprity#options
  settings: {
    retina: true,
    dimension: [
      {
        ratio: 1, dpi: 72
      }, {
        ratio: 2, dpi: 192
      }
    ],
    margin: 0,
    split: true, // to create multiple sprites by putting images in subdirectories
    name: 'sprite', // for split. ex: sprite-main.png, sprite-blog.png
    style: '_sprites.scss',
    cssPath: p.join(config.scaffold.assets.root, config.scaffold.assets.styles),
    template: p.join(config.scaffold.source.root, config.scaffold.source.sprites, '/template/scss.hbs'),
    processor: 'sass',
    orientation: 'binary-tree',
    prefix: 'sprite' // for sass
  }
}
```

### `browserify`

Compiles all of your CommonJS modules into bundles. Make sure you have configured your bundles.

**Options:**
```javascript
options.browserify = {
  debug: false,
  // path to the config file where you will set up your bundles
  bundleConfig: '_src/js/config/bundles.js',
  // to set up bundles directly in your options
  // not recommended because using this method will not
  // automatically bundle when using the `watch` task
  bundles: [],
  transform: ['browserify-shim', 'browserify-ejs'],
  aliases: { "waitFor": p.join(source.scripts, "/lib/waitFor.js") },
  shim: { "jquery": "global:$" }
}
```

### `static`

Creates static HTML files from HTML partials

**Options:**
```javascript
options.static = {
  extension: ".html",
  // For all available settings: https://www.npmjs.com/package/gulp-file-include
  settings: {
    prefix: '@@',
    basepath: '@file'
  }
}
```

### `watch`

Watches for changes on source files, and when a file is added, removed, or edited, it runs the necessary task.

**Options:**
```javascript
// key: the source folder to watch for changes on
// value: the task to run when a change happens
config.watch = {
  scripts:  'browserify',
  styles:   'sass',
  symbols:  'symbols',
  images:   'images',
  sprites:  'sprites',
  static:   'static'
};
```

**To add a new dir to watch:**

1. Add it to `scaffold.source`
3. Add a new glob for the files in that folder
2. Add the folder in the `watch` config along with the task you want to run when changes are made.

*Example:*
```javascript
eta(gulp, {
  scaffold: {
    source: {
      custom: 'my-custom-dir'
    }
  },
  globs: {
    custom: '**/*.+(txt)'
  },
  watch: {
    custom: 'my-custom-task'
  }
});

gulp.task('my-custom-task', function() {
  console.log('this task is super awesome');
});
```

### `browserSync`

Starts Browser Sync and runs `watch` in tandem.

**Options:**
```javascript
config.browserSync = {
  useBrowserSync: true,
  // For all available settings: http://www.browsersync.io/docs/options/
  settings: {
    server: appDir,
    open: true,
    notify: false,
    reloadOnRestart: true,
    files: [
      p.join(assets.root, '**/*'),
      p.join(appDir, '**/*.php'),
    ]
  }
}
```

### `minifyCss`

Minifies your compiled stylesheets

### `minifyJs`

Minifies your compiled JavaScript files

### `production`

Re-builds optimized, compressed css and js files to the assets folder, as well as output their file sizes to the console. It's a shortcut for running the following tasks: `['minifyCss', 'uglifyJs']`.


CHANGELOG.md
## Overall

* Convert to NPM. You can find the package on [NPM](https://www.npmjs.com/package/gulp-eta)
* Added in BrowserSync support & removed LiveReload
* Sass globbing does not work due to bug in libsass update
* Updated sprite generation to a new package. Crisp sprites now!
* Updated `index.html` sample with wysiwyg, symbols, & sprite usage
* Removed Media Query Combine as it wasn't combining media queries correctly
* Broke out the config file into individual modules for better organization

## Browserify

**Changes**
* Item

**Added**
* Item

**Removed**
* Item

## Sass

**Changes**
* Converted remaining px values to rem
* `/app/` is now `/pages/`
* `/helpers/` is now `/util/`
* Moved clearfix helpers from `/base/reset.scss` to `/utils/utilities.scss`

**Added**
* Mixin to manage font-size & line-height `/util/mixins/type.scss`
* `/components/` folder to handle smaller widget style modules
* Added in additional defaults for print styles

**Removed**
* Removed wysiwyg styles export to WP admin. (styling & admin CP issues)
* `/vendor/` should only contain a sample file now
* Removed all mentions of fontello as we have our own symbol font
* Removed `/layout/sidebar.scss` as it wasn't being utilizied

## Sprites

**Changes**
* You can have multiple sprite sets now. Simply organize sprites through folders in `/_src/sprites/`
* New file names would be `sprite-folder_name`
* You still have to `@extend .sprite-folder_name` to pull in correct sprite


lib/config.js
var p = require('path');
var merge = require('./util/deepishMerge.js');

// Setup objects
var source    = {};
var assets    = {};
var srcs      = {};
var config    = {};

var _assets = function(asset) {
	return config.scaffold.assets[asset].charAt(0) === '/'
		? p.join(appDir, config.scaffold.assets[asset])
		: p.join(assets.root, config.scaffold.assets[asset]);
};

// Gulp options/settings for tasks.
module.exports = function(base, options) {

	// Paths
	appDir = base;

	// Globs
	config.globs = {
		images:   '**/*.+(png|jpg|jpeg|svg|gif)',
		scripts:  '**/*.+(js|ejs)',
		sprites:  '**/*.+(png)',
		styles:   '**/*.+(sass|scss)',
		symbols:  '*.+(svg)',
		static:   '**/*.+(html)'
	};

	// Scaffold folders
	config.scaffold = {
		source: {
			root:     '_src',
			images:   'images',
			scripts:  'js',
			sprites:  'sprites',
			styles:   'sass',
			symbols:  'symbols',
			static:   'static'
		},
		assets: {
			root:     'assets',
			images:   'images',
			sprites:  'images/sprites',
			scripts:  'js',
			styles:   '/',
			symbols:  'fonts/symbols',
			static:   '/'
		}
	};

	// merge the user scaffold with the defaults
	if (options.scaffold) {
		config.scaffold = merge(config.scaffold, options.scaffold);
	}

	// Source Directory
	source.root     = p.join(appDir, config.scaffold.source.root);
	source.images   = p.join(source.root, config.scaffold.source.images);
	source.scripts  = p.join(source.root, config.scaffold.source.scripts);
	source.sprites  = p.join(source.root, config.scaffold.source.sprites);
	source.styles   = p.join(source.root, config.scaffold.source.styles);
	source.symbols  = p.join(source.root, config.scaffold.source.symbols);
	source.static   = p.join(source.root, config.scaffold.source.static);
	source.symTpl   = p.join(source.symbols, 'template');

	// Assets Directory
	assets.root     = p.join(appDir, config.scaffold.assets.root);
	assets.images   = _assets('images');
	assets.scripts  = _assets('scripts');
	assets.sprites  = _assets('sprites');
	assets.styles   = _assets('styles');
	assets.symbols  = _assets('symbols');
	assets.static   = _assets('static');

  // Srcs for gulp.src
  srcs.images     = p.join(source.images, config.globs.images);
  srcs.scripts    = p.join(source.scripts, config.globs.scripts);
  srcs.sprites    = p.join(source.sprites, config.globs.sprites);
  srcs.styles     = p.join(source.styles, config.globs.styles);
  srcs.symbols    = p.join(source.symbols, config.globs.symbols);
  srcs.static     = p.join(source.static, '*.tpl.html');

	// Images
	config.images = {
		settings: {
			progressive: true,
			optimizationLevel: 4
		}
	};

	// Browserify
	config.browserify = {
		debug: false,
		bundles: [],
		transform: ['browserify-shim', 'browserify-ejs'],
		aliases: { "waitFor": './' + p.join(config.scaffold.source.root, config.scaffold.source.scripts, "/lib/waitFor.js") },
		shim: { "jquery": "global:$" }
	};

	// Sprites
	config.sprites = {
		destSass: p.join(source.styles, 'utils'),
		settings: {
			src: srcs.sprites,
			retina: true,
			dimension: [
				{
	        ratio: 1, dpi: 72
	      }, {
	        ratio: 2, dpi: 192
	      }
      ],
			margin: 0,
			split: true, // to create multiple sprites by putting images in subdirectories
			name: 'sprite', // for split. ex: sprite-main.png, sprite-blog.png
			style: '_sprites.scss',
			cssPath: p.join(config.scaffold.assets.root, config.scaffold.assets.sprites),
			template: p.join(config.scaffold.source.root, config.scaffold.source.sprites, '/template/scss.hbs'),
			processor: 'sass',
			orientation: 'binary-tree',
			prefix: 'sprite' // for sass
		}
	};

	// Sass
	config.sass = {
		settings: {
			indentedSyntax: true,
			errLogToConsole: true,
			outputStyle: 'nested'
		},
		globbing: {
			extensions: ['.scss', '.sass']
		}
	};

	// Auto Prefixer
	config.autoprefixer = {
		browsers: [
			'last 2 versions',
			'safari 5',
			'ie 8',
			'ie 9',
			'android 4'
		],
		cascade: true
	};

	// Symbols
	config.symbols = {
		tplCss: p.join(source.symTpl, 'symbols.tpl.css'),
		tplSass: p.join(source.symTpl, 'symbols.tpl.scss'),
		tplHtml: p.join(source.symTpl, 'symbols.tpl.html'),
		fontPath: p.join(config.scaffold.assets.root, config.scaffold.assets.symbols, '/'),
		destSass: p.join(source.styles, 'utils'),
		settings: {
			fontName: 'symbols',
			appendCodepoints: false,
			centerHorizontally: true,
			normalize: true,
			fontHeight: false
		},
		renameSass: {
			basename: '_symbols',
			extname: '.scss'
		}
	};

	// Static
	config.static = {
		extension: ".html",
		// see https://www.npmjs.com/package/gulp-file-include for available settings
		settings: {
			prefix: '@@',
			basepath: '@file'
		},
	};

	// BrowserSync
	config.browserSync = {
		useBrowserSync: true,
		// See http://www.browsersync.io/docs/options/ for a complete list of configuration options
		settings: {
			server: appDir,
			open: true,
			notify: false,
			reloadOnRestart: true,
			files: [
				p.join(assets.root, '**/*'),
				p.join(appDir, '**/*.php'),
			]
		}
	};

	// Default Tasks
	config.default = {
		tasks: ['browserSync', 'symbols', 'sass', 'sprites', 'images', 'browserify']
	};

	// Watch
	config.watch = {
		scripts:  'browserify',
		styles:   'sass',
		symbols:  'symbols',
		images:   'images',
		sprites:  'sprites',
		static:   'static'
	};

	config.init = {
		srcDir: p.join(p.dirname(module.filename),'_src'),
		cwd: appDir,
		manifestPath: p.join(appDir, 'package.json'),
		gulpfile: p.join(appDir, 'gulpfile.js'),
		bundlesPath: './' + config.scaffold.source.root + '/' + config.scaffold.source.scripts + '/config/bundles.js',
		dependencies: ['sassqwatch'],
		devDependencies: ['gulp', 'browserify-ejs', 'browserify-shim'],
		stuffToAppend: {
			'browserify': {
				'transform': config.browserify.transform,
			},
			'browser': config.browserify.aliases,
			'browserify-shim': config.browserify.shim
		}
	};

	// merge the default config with the custom settings
	config = merge(config, options);

	// expose the scaffold and paths after the merge
	// so that it can't be overridden
	config.source    = source;
	config.assets    = assets;
	config.srcs      = srcs;

	return config;
};